phoenix-core-server/src/main/java/org/apache/phoenix/replication/ReplicationLog.java

```diff
@@ -0,0 +1,857 @@
+        Path shardPath = new Path(url.getPath(), String.format(SHARD_DIR_FORMAT, shard));
+        // Ensure the shard directory exists. We track which shard directories we have probed or
+        // created to avoid a round trip to the namenode for repeats.
+        IOException[] exception = new IOException[1];
+        shardMap.computeIfAbsent(shardPath, p -> {
+            try {
+                if (!fs.exists(p)) {
+                    if (!fs.mkdirs(p)) {
+                        throw new IOException("Could not create path: " + p);
+                    }
+                }
+            } catch (IOException e) {
+                exception[0] = e;
+            }
+            return p;
+        });
+        // If we faced an exception in computeIfAbsent, throw it
+        if (exception[0] != null) {
+            throw exception[0];
+        }
+        Path filePath = new Path(shardPath, String.format(FILE_NAME_FORMAT, timestamp, serverName));
+        return filePath;
+    }
+
+    /**
+     * Creates and initializes a new LogFileWriter.
+     * @param fs  The FileSystem to write to.
+     * @param url A URI pointing to the root log directory.
+     * @return The newly created and initialized LogFileWriter.
+     * @throws IOException If an error occurs during file creation or writer initialization.
+     */
+    protected LogFileWriter createNewWriter(FileSystem fs, URI url) throws IOException {
+        Path filePath = makeWriterPath(fs, url);
+        LogFileWriterContext writerContext = new LogFileWriterContext(conf).setFileSystem(fs)
+            .setFilePath(filePath).setCompression(compression);
+        LogFileWriter newWriter = new LogFileWriter();
+        try {
+            newWriter.init(writerContext);
+            newWriter.setGeneration(writerGeneration.incrementAndGet());
+        } catch (IOException e) {
+            LOG.error("Failed to initialize new LogFileWriter for path {}", filePath, e);
+            throw e;
+        }
+        return newWriter;
+    }
+
+    protected void closeWriter(LogFileWriter writer) {
+        if (writer == null) {
+            return;
+        }
+        try {
+            writer.close();
+        } catch (IOException e) {
+            // For now, just log and continue
+            LOG.error("Error closing log writer: " + writer, e);
+        }
+    }
+
+    /** Force closes the log upon an unrecoverable internal error. */
+    protected void closeOnError() {
+        // Stop the time based rotation check.
+        stopRotationExecutor();
+        // We expect a final sync will not work. Just close the inner writer.
+        if (currentWriter != null) {
+            closeWriter(currentWriter);
+            currentWriter = null;
+        }
+        // Directly halt the disruptor. shutdown() would wait for events to drain. We are expecting
+        // that will not work.
+        disruptor.halt();
+    }
+
+    /** Closes the log. */
+    public void close() {
```

Comment:
It seems like this method is not threadsafe.

Reply:
Addressed in a5a81cecd